configfile: "config.yaml"

# TODO: aggregate metadata on kallisto runs in df and choose version of callisto used for most of the samples.
# TODO: gz all tsv files in pipeline

# TODO: add QC based on variance. Something strange with GSE75171
# TODO: IMPLIMENT PRINCIPAL HULL ANALYSIS WITH ELBOW THRESHOLDING
# TODO: Meaningfuly scale loadings for weighted gsea


import pandas as pd
from os import listdir
from os.path import isfile, join
import re

glob_gsm_srr_map = {}
glob_gse_gsm_map = {}
glob_postquant_gse_gsm_map = {}

rule all:
    input:
        expand("out/{organism}/{platform}/pca_ks/"
               "{n_genes}_{scale}_{max_comp}_{var_threshold}/"
               "prepared/{geneset_name}.tsv",
                organism='mm',
                platform=['chip'],
                n_genes="6000",
                scale="linear",
                max_comp="10",
                var_threshold="0.02",
                geneset_name=["HALLMARK_HYPOXIA", "GSE120744_TREM_SIGNATURE"])
        # "out/mm/chip/exp_mat_qc.tsv"
        # "out/mm/chip/pca/6000_linear/GSE75171.rds"
        #"out/mm/chip/pca/6000_linear/GSE8150.rds"
        # "out/mm/seq/gse_cpm_qc.tsv"
        # "out/mm/seq/pca/3000_log_10_0.02_PCList.rds"
        #"out/mm/seq/gsm_qc.tsv"
        # "out/mm/seq/gses/GSE100000.tsv"
        # expand("out/hs/seq/gsms/{gsm}.tsv", gsm=["GSM3739518","GSM3739800","GSM3740121","GSM3740090","GSM3740162","GSM3740326","GSM3739972","GSM3872712","GSM3560575","GSM3872850","GSM3739649","GSM3739844","GSM3740331","GSM3740434","GSM3740516","GSM3633773","GSM3633873","GSM3740066","GSM3740511","GSM1241207","GSM3740690","GSM3740742","GSM1241137","GSM3739542","GSM3739315","GSM3633877","GSM3872857","GSM3739696","GSM3740430","GSM3739973","GSM3560532","GSM1708824","GSM3740040","GSM3740722","GSM3739729","GSM3740421","GSM3851919","GSM3740696","GSM3633696","GSM3872680","GSM3740490","GSM3740661","GSM3740113","GSM3735404","GSM3740439","GSM3740095","GSM3872818","GSM3560570","GSM3739705","GSM3740253","GSM3740230","GSM3851899","GSM3739839","GSM3739480","GSM3739925","GSM3739836","GSM3633723","GSM3872827","GSM3851890","GSM3560633","GSM3739341","GSM3740561","GSM3739861","GSM3560632","GSM3739357","GSM3633819","GSM3735425","GSM3739938","GSM3739746","GSM3739909","GSM3872806","GSM3735308","GSM3560594","GSM3851904","GSM3739827","GSM3739605","GSM3851879","GSM3740723","GSM3633874","GSM3633906","GSM3740182","GSM3740486","GSM3740309","GSM3739405","GSM3740164","GSM3740632","GSM3851931","GSM3872813","GSM3740086","GSM3740045","GSM3740316","GSM3740501","GSM3851917","GSM3740475","GSM3740311","GSM3633695","GSM3735370","GSM3633712","GSM3739387","GSM3740568","GSM3740271","GSM3740415","GSM3740549","GSM3739543","GSM3739375","GSM3740618","GSM3740148","GSM3740541","GSM3740744","GSM3739600","GSM3739330","GSM3739321","GSM3740536","GSM3740396","GSM3739895","GSM3740405","GSM3740534","GSM3740677","GSM3560552","GSM3740124","GSM3740251","GSM3872809","GSM3739347","GSM3740728","GSM3633910","GSM3740586","GSM3560523","GSM3735343","GSM3740498","GSM3851822","GSM3851848","GSM3851867","GSM3740551","GSM3872856","GSM3740724","GSM3735362","GSM3740665","GSM3740115","GSM3633697","GSM3740234","GSM1241167","GSM3735371","GSM3633791","GSM3633822","GSM3633805","GSM3633894","GSM3740388","GSM3740605","GSM3739889","GSM3739575","GSM3633768","GSM3872665","GSM3739597","GSM3740293","GSM3739418","GSM1241138","GSM3739463","GSM3739984","GSM3872893","GSM3740535","GSM3735441","GSM1708821","GSM3633843","GSM3739667","GSM3739930","GSM1241177","GSM3568643","GSM3739669","GSM3735384","GSM3633750","GSM3740076","GSM3872884","GSM3735356","GSM3851891","GSM3740770","GSM3739739","GSM3633756","GSM3872910","GSM3560572","GSM3740122","GSM3740049","GSM3740703","GSM3740218","GSM3740422","GSM3735422","GSM3872719","GSM3740291","GSM3740333","GSM3851929","GSM3633759","GSM3739697","GSM3851831","GSM3740126","GSM3740687","GSM3872879","GSM3739494","GSM3740233","GSM3740642","GSM3739699","GSM3851918","GSM3739670","GSM3740034","GSM3740408","GSM3739849","GSM3735414","GSM3739868","GSM3740582","GSM3739325","GSM3739515","GSM3739919","GSM3735347","GSM3872769","GSM3740135","GSM3740451","GSM3740380","GSM3740735","GSM3560592","GSM3872898","GSM3633719","GSM3739978","GSM3739717","GSM3739538","GSM3740167","GSM3739716","GSM3739830","GSM3872845","GSM3735432","GSM3739790","GSM3851844","GSM3740029","GSM3851820","GSM3739445","GSM3740200","GSM3740753","GSM3740217","GSM3740626","GSM3739618","GSM1708822","GSM3739884","GSM3872758","GSM3739860","GSM3740371","GSM3740319","GSM3633703","GSM3740004","GSM3872783","GSM3740611","GSM3740727","GSM3872852","GSM3740337","GSM3851896","GSM3740467","GSM3739850","GSM3739963","GSM3560600","GSM3740110","GSM3735329","GSM3851924","GSM3740145","GSM3633743","GSM3739801","GSM3739683","GSM3740185","GSM3740231","GSM3740548","GSM3740186","GSM3740292","GSM3739685","GSM3560543","GSM3740069","GSM3560521","GSM3735323","GSM3851940","GSM3739712","GSM3740084","GSM3740063","GSM3740721","GSM3739992","GSM3633713","GSM3872781","GSM3735330","GSM3633767","GSM3633818","GSM3740645","GSM3740147","GSM3739462","GSM3740083","GSM3739447","GSM3739572","GSM3739554","GSM3851897","GSM3560579","GSM3740056","GSM3739509","GSM3735361","GSM1241180","GSM3633747","GSM3740193","GSM3739583","GSM3739390","GSM3633776","GSM3560559","GSM3739714","GSM3560540","GSM3739911","GSM3740485","GSM3633845","GSM3872877","GSM3740608","GSM3739706","GSM3872690","GSM3739578","GSM3633708","GSM3633838","GSM1241174","GSM3872823","GSM1241241","GSM3740047","GSM3740565","GSM3740359","GSM3740617","GSM3739872","GSM3740000","GSM3872701","GSM3740112","GSM3739791","GSM3735426","GSM3740142","GSM3872683","GSM3740588","GSM3739468","GSM3740285","GSM3739964","GSM3560619","GSM3735309","GSM3735376","GSM3560599","GSM3872764","GSM3740387","GSM3872773","GSM3851893","GSM3740429","GSM3851942","GSM3872766","GSM3740643","GSM3739887","GSM3740745","GSM3633825","GSM3740553","GSM3735427","GSM3739945","GSM3739601","GSM3740589","GSM3740506","GSM3739424","GSM3740279","GSM3740295","GSM3740545","GSM3740223","GSM3740289","GSM3740623","GSM3740689","GSM3735383","GSM1241192","GSM3633827","GSM3633736","GSM3568642","GSM3633729","GSM3740208","GSM3740426","GSM3740381","GSM3560637","GSM3851884","GSM3740317","GSM1241201","GSM3740575","GSM3740504","GSM3739638","GSM3739611","GSM3872710","GSM3735336","GSM3739425","GSM3740635","GSM3740298","GSM3740746","GSM3739437","GSM3739379","GSM3740227","GSM3739745","GSM1241183","GSM1241200","GSM3740144","GSM3740094","GSM3739503","GSM3739563","GSM3872860","GSM3740050","GSM3739497","GSM3740471","GSM3560612","GSM3560538","GSM3740341","GSM3560601","GSM3739952","GSM3740367","GSM3872653","GSM1241204","GSM3851864","GSM3739835","GSM3851886","GSM3740003","GSM3740507","GSM3633826","GSM3872897","GSM3560528","GSM3739693","GSM3872866","GSM3739762","GSM3739937","GSM3872679","GSM3739873","GSM3740274","GSM3739589","GSM3739340","GSM3739345","GSM3560547","GSM3740238","GSM3739541","GSM1241230","GSM3872849","GSM3851914","GSM1241236","GSM3739523","GSM3740366","GSM3740256","GSM3633760","GSM1241210","GSM3872761","GSM3739348","GSM3740758","GSM3633809","GSM3739995","GSM3633913","GSM1241165","GSM1241219","GSM3739629","GSM3740214","GSM3872914","GSM3633836","GSM3739734","GSM3633745","GSM3872825","GSM3739449","GSM3872907","GSM3851854","GSM3739657","GSM3740676","GSM3851926","GSM3872812","GSM3740425","GSM3739818","GSM3739766","GSM3560589","GSM3739773","GSM3872855","GSM3633830","GSM3739648","GSM3739370","GSM3739478","GSM3872703","GSM3560631","GSM3740609","GSM3739485","GSM3740754","GSM3872709","GSM3739410","GSM3633746","GSM3633857","GSM3872798","GSM3740344","GSM3739968","GSM3740521","GSM3740312","GSM3739647","GSM3872819","GSM3739941","GSM3739970","GSM1241217","GSM3739863","GSM3740492","GSM3739767","GSM3740629","GSM3633871","GSM3851853","GSM3872744","GSM3740272","GSM3739845","GSM3739327","GSM3739829","GSM3735430","GSM3560556","GSM3560593","GSM3739723","GSM3739619","GSM1241159","GSM3740591","GSM3740052","GSM3872763","GSM1241246","GSM3740091","GSM3740021","GSM3872883","GSM3739689","GSM3739736","GSM3740557","GSM3851927","GSM3739331","GSM3739402","GSM3872681","GSM3740013","GSM3740413","GSM3851894","GSM3560614","GSM3739384","GSM3740192","GSM3633902","GSM3740061","GSM3740370","GSM3739775","GSM1241154","GSM3740525","GSM3739742","GSM3739748","GSM3739959","GSM3739407","GSM3735413","GSM3739983","GSM3739838","GSM3872789","GSM3739752","GSM3740500","GSM1241244","GSM3560636","GSM3739429","GSM3740132","GSM3739511","GSM3740136","GSM3740747","GSM3872787","GSM1241176","GSM3560587","GSM3740108","GSM3633909","GSM3735387","GSM3740600","GSM3872890","GSM3872749","GSM3739628","GSM3739632","GSM3740474","GSM3740268","GSM3740670","GSM3740539","GSM1241169","GSM3872685","GSM3740157","GSM3872691","GSM3872677","GSM3872822","GSM3740259","GSM3739751","GSM3740358","GSM3739625","GSM3740443","GSM3851877","GSM3739765","GSM3740496","GSM1708815","GSM3872729","GSM3872750","GSM3872715","GSM3872725","GSM3739361","GSM3740466","GSM3740125","GSM3560586","GSM3739334","GSM3740365","GSM3872788","GSM3740022","GSM3560595","GSM3740346","GSM3851814","GSM3739636","GSM3633861","GSM3739780","GSM3740379","GSM3740449","GSM3851938","GSM3740024","GSM3740647","GSM3740324","GSM3735352","GSM3740105","GSM3560616","GSM3740660","GSM3735333","GSM3739672","GSM3740522","GSM3740736","GSM3740027","GSM3872851","GSM3739421","GSM3872899","GSM3872908","GSM3740243","GSM3739897","GSM3633863","GSM3872847","GSM3851900","GSM3739397","GSM3740584","GSM1241206","GSM3739547","GSM3740615","GSM3739738","GSM3740246","GSM3740580","GSM3739366","GSM3740111","GSM3560533","GSM3739529","GSM1241242","GSM3633717","GSM3740235","GSM3560573","GSM3740282","GSM3872824","GSM3740576","GSM3740321","GSM3872720","GSM3739703","GSM3740627","GSM3872846","GSM3739481","GSM3740060","GSM3560598","GSM3872811","GSM3739908","GSM3739936","GSM3740252","GSM3740477","GSM3633701","GSM3872782","GSM3739309","GSM3633808","GSM3740043","GSM3740752","GSM3739482","GSM3851869","GSM3740266","GSM3872796","GSM1241157","GSM3740099","GSM3560548","GSM1241178","GSM3740130","GSM3739313","GSM3739808","GSM3633761","GSM3633722","GSM3633781","GSM3735335","GSM3739824","GSM3872815","GSM3739890","GSM3633848","GSM3560555","GSM3739471","GSM3740196","GSM3740068","GSM3739969","GSM3633744","GSM3739642","GSM3739537","GSM3739948","GSM3739412","GSM3633784","GSM3740457","GSM3740201","GSM3735307","GSM3740692","GSM3735315","GSM3739659","GSM3740682","GSM3740624","GSM3851862","GSM3739795","GSM3740165","GSM3740563","GSM3740734","GSM3739981","GSM3740288","GSM3735369","GSM3739566","GSM3739304","GSM3740078","GSM3851858","GSM3740221","GSM3735306","GSM3735381","GSM3739506","GSM3740564","GSM3735332","GSM3872791","GSM3851874","GSM3872666","GSM3739565","GSM3872664","GSM3740717","GSM3740556","GSM3740581","GSM3740212","GSM3739788","GSM3739691","GSM3851921","GSM3735354","GSM3851829","GSM3739931","GSM3739403","GSM3633804","GSM3560564","GSM3739977","GSM3739869","GSM3740685","GSM3633847","GSM3740532","GSM3740653","GSM3739879","GSM3740505","GSM3739812","GSM3735385","GSM3739386","GSM3633775","GSM3740375","GSM3872900","GSM3739803","GSM3740313","GSM3872670","GSM1241228","GSM3633875","GSM3633800","GSM3633699","GSM3739737","GSM3560550","GSM3739658","GSM3633770","GSM3872810","GSM1708823","GSM3872804","GSM3872673","GSM3739815","GSM3740071","GSM3735355","GSM3851826","GSM3739764","GSM3740002","GSM3740123","GSM3740666","GSM3872732","GSM1241184","GSM3740158","GSM3633787","GSM3872743","GSM3872695","GSM3740265","GSM3739652","GSM3739544","GSM3739867","GSM3872834","GSM3739546","GSM3739997","GSM3740156","GSM3740332","GSM3740222","GSM3740527","GSM3560602","GSM3740340","GSM3851856","GSM3872779","GSM3740149","GSM3739637","GSM3740140","GSM3560580","GSM3740262","GSM3740633","GSM3633844","GSM3740383","GSM3740663","GSM3740641","GSM3740448","GSM3560585","GSM3740224","GSM3872723","GSM3739593","GSM3851834","GSM3739595","GSM3633868","GSM3740574","GSM3739662","GSM3740195","GSM3739965","GSM3739570","GSM3740648","GSM3633869","GSM3739326","GSM3739816","GSM3740249","GSM3739470","GSM3739476","GSM3633780","GSM3739907","GSM3740219","GSM3740294","GSM3739951","GSM3560563","GSM3740241","GSM3872844","GSM3739774","GSM3560561","GSM3633856","GSM3739533","GSM3872661","GSM3739871","GSM3560608","GSM3851887","GSM3872839","GSM3739974","GSM3739483","GSM3739486","GSM3735351","GSM3740131","GSM3739490","GSM3735444","GSM3851824","GSM3633749","GSM3739423","GSM3740693","GSM3739491","GSM3740006","GSM3560574","GSM3633751","GSM3739427","GSM3739630","GSM3740261","GSM3739376","GSM3560623","GSM3740242","GSM3739730","GSM3560560","GSM3851906","GSM3740741","GSM3739851","GSM3633813","GSM3633897","GSM3851911","GSM3735337","GSM3739976","GSM3740649","GSM3740390","GSM3739666","GSM3740376","GSM3740542","GSM3740547","GSM3735360","GSM3739461","GSM3739686","GSM3633870","GSM3739507","GSM3739918","GSM3739720","GSM3851935","GSM1241172","GSM3740152","GSM3740382","GSM3740593","GSM3740016","GSM3872901","GSM3872759","GSM3739854","GSM3740250","GSM3735377","GSM3851912","GSM3740278","GSM3872797","GSM3633824","GSM3739796","GSM3851889","GSM3560611","GSM3560530","GSM3740655","GSM3739927","GSM3739603","GSM3740389","GSM3735349","GSM3739553","GSM3740438","GSM3740461","GSM3739487","GSM3740654","GSM3739335","GSM3740512","GSM3740161","GSM3739655","GSM3872868","GSM3633852","GSM1241198","GSM3740213","GSM3739677","GSM3740058","GSM3739579","GSM3739474","GSM3740228","GSM3735320","GSM3740042","GSM3872736","GSM3560567","GSM3735423","GSM3739534","GSM3560571","GSM3740705","GSM3739360","GSM3740065","GSM3740315","GSM3872687","GSM3740650","GSM1241164","GSM3633912","GSM3633810","GSM3633880","GSM3739813","GSM3740695","GSM3739489","GSM3740640","GSM3740651","GSM3740290","GSM3633898","GSM3740699","GSM3739520","GSM3740297","GSM3633704","GSM3739832","GSM3740206","GSM3739373","GSM3740360","GSM3740571","GSM3740220","GSM3740395","GSM3739892","GSM3735328","GSM3740737","GSM3739560","GSM3739401","GSM3739953","GSM3872837","GSM3739881","GSM3740740","GSM3633850","GSM3739810","GSM3740694","GSM3739359","GSM3739718","GSM3739337","GSM3735342","GSM3633917","GSM3735378","GSM3740137","GSM3735438","GSM1241186","GSM3633878","GSM3740046","GSM3739822","GSM3560618","GSM3739663","GSM3740619","GSM3633757","GSM3872915","GSM3735340","GSM3735339","GSM3739329","GSM1708829","GSM3740463","GSM3740318","GSM3739859","GSM3560520","GSM3633766","GSM3739477","GSM3739750","GSM3740118","GSM3560534","GSM3739673","GSM3633725","GSM3739298","GSM3633710","GSM3872762","GSM3740765","GSM3740526","GSM3872904","GSM3740184","GSM3739413","GSM1241160","GSM3740339","GSM3739875","GSM3739674","GSM3872906","GSM3560606","GSM3633858","GSM3739499","GSM3872713","GSM1241181","GSM3560604","GSM3560620","GSM3560537","GSM3739719","GSM3633882","GSM3740707","GSM3560576","GSM3633867","GSM3739947","GSM3740590","GSM3739893","GSM3633829","GSM3872865","GSM3633799","GSM3739306","GSM3739498","GSM3872776","GSM3740460","GSM3735405","GSM3739562","GSM3740336","GSM3740569","GSM3740305","GSM3633715","GSM3740767","GSM3740378","GSM3740310","GSM3740356","GSM3739819","GSM3740625","GSM3633888","GSM3872705","GSM3872652","GSM3740320","GSM3872778","GSM3740488","GSM3872795","GSM3740432","GSM1241243","GSM3740211","GSM3560607","GSM3739596","GSM3739433","GSM3739316","GSM3740308","GSM3740732","GSM1241202","GSM3740203","GSM3740102","GSM3735409","GSM3740634","GSM3739656","GSM3740127","GSM3872870","GSM3872671","GSM3633782","GSM1708827","GSM3739599","GSM3872658","GSM3739302","GSM3740107","GSM3740391","GSM3872735","GSM3560518","GSM3739707","GSM1241229","GSM3560565","GSM3740552","GSM3872718","GSM3872801","GSM3739469","GSM3739364","GSM3740364","GSM3851871","GSM3739557","GSM3739990","GSM3740673","GSM3872657","GSM3740566","GSM3633833","GSM3740191","GSM1241227","GSM3740304","GSM3740599","GSM3739668","GSM3740598","GSM3872784","GSM3851810","GSM3739414","GSM3739865","GSM3740064","GSM3740187","GSM3740354","GSM3739799","GSM3872874","GSM3740678","GSM3740385","GSM3739678","GSM3739704","GSM3739848","GSM3633709","GSM3739440","GSM3740464","GSM3560562","GSM3740146","GSM3735334","GSM3740436","GSM3740470","GSM3739906","GSM3740414","GSM3740053","GSM3740679","GSM3740495","GSM3740407","GSM3740603","GSM3739749","GSM3739690","GSM3633889","GSM3739756","GSM3739584","GSM3740079","GSM3739779","GSM3739617","GSM3739958","GSM3740639","GSM3740594","GSM3740560","GSM3560621","GSM3739430","GSM3735327","GSM3739500","GSM3739754","GSM3872803","GSM3872696","GSM3739311","GSM3739301","GSM3739913","GSM3740437","GSM3872790","GSM3872853","GSM3739772","GSM3740749","GSM3740756","GSM1241245","GSM3739368","GSM3560545","GSM3740592","GSM1241173","GSM3739460","GSM3560529","GSM3740352","GSM3740442","GSM3740668","GSM3560638","GSM3872785","GSM3740529","GSM3739303","GSM3633806","GSM3739422","GSM3740348","GSM3851916","GSM3735313","GSM1241239","GSM3739297","GSM3740473","GSM1241218","GSM3740104","GSM3739725","GSM3739508","GSM3560549","GSM3740174","GSM3740621","GSM1241238","GSM3739679","GSM3735429","GSM3740399","GSM3735424","GSM3735398","GSM3633762","GSM3633820","GSM3872835","GSM3872880","GSM3739349","GSM3633895","GSM3740631","GSM3568644","GSM3739442","GSM3739771","GSM3739404","GSM3872905","GSM3872882","GSM3739809","GSM3735442","GSM3740755","GSM1241212","GSM3740398","GSM3735344","GSM3740166","GSM3633779","GSM3740483","GSM3740446","GSM3872754","GSM3740073","GSM3740636","GSM3740578","GSM3740392","GSM3633883","GSM3872686","GSM3735312","GSM3739351","GSM3740057","GSM3740119","GSM3633764","GSM3739446","GSM3735417","GSM3851837","GSM3739519","GSM3872699","GSM3739935","GSM3739609","GSM3740672","GSM3735439","GSM3872727","GSM3740210","GSM3633769","GSM3851835","GSM3633811","GSM3740368","GSM3739682","GSM3740180","GSM3739828","GSM3739759","GSM3560531","GSM3872799","GSM3740072","GSM3740750","GSM3739551","GSM3740323","GSM3739453","GSM3633908","GSM3872694","GSM3740431","GSM3740280","GSM3739942","GSM3633860","GSM3739377","GSM3740005","GSM3633772","GSM3740018","GSM3872828","GSM3740469","GSM3740637","GSM3740597","GSM3739664","GSM3735390","GSM3739342","GSM3872836","GSM3872730","GSM3740503","GSM3739994","GSM3735421","GSM3739920","GSM3735410","GSM3872675","GSM3740269","GSM1708828","GSM3872912","GSM3740684","GSM3740067","GSM3872814","GSM3739545","GSM3740237","GSM3739962","GSM3872774","GSM3872756","GSM3633783","GSM3851830","GSM3739650","GSM3739514","GSM1241193","GSM3851860","GSM3739344","GSM3633890","GSM3739525","GSM3740300","GSM3739531","GSM3740277","GSM3739535","GSM3872792","GSM3739769","GSM3739621","GSM3740351","GSM3851941","GSM3872838","GSM3851861","GSM3739744","GSM3851907","GSM3740197","GSM3633700","GSM3739352","GSM3735341","GSM3739455","GSM3633896","GSM3735310","GSM3872669","GSM3851909","GSM3633740","GSM3872889","GSM3633915","GSM3740739","GSM3739914","GSM3872702","GSM3633866","GSM1241140","GSM3740011","GSM3740484","GSM3560535","GSM3560546","GSM3740419","GSM3872733","GSM3740528","GSM3872767","GSM3740209","GSM3872747","GSM3560609","GSM3851821","GSM3851825","GSM3739654","GSM3739653","GSM3851943","GSM3739807","GSM3633859","GSM1708830","GSM3633834","GSM3740153","GSM3633855","GSM1241155","GSM3740239","GSM3739307","GSM3739874","GSM3740738","GSM3872854","GSM3740606","GSM3739392","GSM3739877","GSM3740299","GSM3851849","GSM3735353","GSM1241179","GSM3560584","GSM3739350","GSM3740698","GSM3740731","GSM3633730","GSM3740423","GSM3740480","GSM3633795","GSM3633789","GSM3739385","GSM3735436","GSM3739841","GSM1241249","GSM3851818","GSM3740190","GSM3633899","GSM3735412","GSM3633796","GSM3739467","GSM3740725","GSM3872808","GSM1708818","GSM3739660","GSM3560553","GSM3735403","GSM3739358","GSM3740244","GSM3740247","GSM3735433","GSM3740445","GSM3735375","GSM3739998","GSM3739821","GSM3568646","GSM1241156","GSM3735367","GSM3739956","GSM3740325","GSM3739431","GSM3739741","GSM3739763","GSM3735418","GSM3740051","GSM3739843","GSM3851885","GSM3740287","GSM3740708","GSM3740342","GSM3739580","GSM3739681","GSM3560577","GSM3735348","GSM3739512","GSM3739880","GSM3739785","GSM3872800","GSM3740493","GSM3740314","GSM3740177","GSM3740260","GSM3633916","GSM3872793","GSM3740411","GSM3872714","GSM3851866","GSM3740303","GSM3740296","GSM3851855","GSM1241247","GSM3739559","GSM3739661","GSM3739797","GSM3740116","GSM3740044","GSM3560583","GSM3740098","GSM3740306","GSM3740499","GSM3740510","GSM3633738","GSM1241168","GSM3740701","GSM3740537","GSM3739758","GSM3560634","GSM3739426","GSM3851888","GSM3739640","GSM3740286","GSM3740751","GSM3633728","GSM3739855","GSM3740129","GSM3739934","GSM3872662","GSM3740585","GSM3740716","GSM3740384","GSM3739823","GSM3740361","GSM3735325","GSM3851920","GSM3735331","GSM3735373","GSM3872753","GSM3740403","GSM3739702","GSM3633837","GSM3560582","GSM3739607","GSM3740759","GSM3735338","GSM3739318","GSM3560626","GSM3739624","GSM3739782","GSM3739408","GSM3739680","GSM3739510","GSM3740281","GSM3740688","GSM3740420","GSM3633821","GSM3633817","GSM3633823","GSM3740404","GSM3633714","GSM3560596","GSM3560522","GSM3739900","GSM3740482","GSM3739415","GSM3740559","GSM3735396","GSM3739760","GSM3740077","GSM3633879","GSM3740462","GSM3739573","GSM1241225","GSM3872771","GSM3739853","GSM3735393","GSM3739826","GSM3735316","GSM3739786","GSM3872832","GSM3872676","GSM3739757","GSM3560568","GSM3740033","GSM3739944","GSM3735364","GSM3739574","GSM1708816","GSM3872903","GSM3739577","GSM3740706","GSM3739635","GSM3740150","GSM3872742","GSM3739328","GSM3740327","GSM3633739","GSM3739856","GSM3739722","GSM3633840","GSM3739967","GSM3740089","GSM3740343","GSM3735346","GSM3740628","GSM3560603","GSM3739319","GSM3739634","GSM3633884","GSM3633832","GSM3739539","GSM1241223","GSM3740509","GSM3740330","GSM3739320","GSM3740338","GSM3872895","GSM3740106","GSM3872708","GSM3739372","GSM3740769","GSM3740026","GSM3735401","GSM3560557","GSM3740519","GSM3739380","GSM3739612","GSM3740604","GSM3739726","GSM3740416","GSM1708819","GSM3560627","GSM3740428","GSM1241235","GSM3633802","GSM3740616","GSM3872888","GSM3872655","GSM3872668","GSM3739501","GSM3633920","GSM3739917","GSM3739694","GSM3740133","GSM3740236","GSM3851898","GSM3739921","GSM3560625","GSM3740307","GSM3633702","GSM3740015","GSM3739980","GSM3740176","GSM1241182","GSM3739820","GSM3740452","GSM3740719","GSM3872875","GSM3560536","GSM3633792","GSM3740141","GSM3740612","GSM3740538","GSM3740644","GSM3872902","GSM3872737","GSM3739866","GSM3568645","GSM3851828","GSM3739382","GSM3739915","GSM3851852","GSM3739986","GSM3739576","GSM3739645","GSM1241171","GSM3740017","GSM3739798","GSM3739885","GSM3633698","GSM3740059","GSM3739389","GSM3633733","GSM3739496","GSM3740550","GSM3740207","GSM3560519","GSM3739949","GSM3740008","GSM3740743","GSM3560635","GSM3735406","GSM3735411","GSM3739710","GSM3740134","GSM3872674","GSM3872833","GSM3872684","GSM3740514","GSM3740533","GSM3560524","GSM3851922","GSM3740151","GSM3872682","GSM3739971","GSM3740757","GSM3739502","GSM3633901","GSM3739305","GSM3851882","GSM3739336","GSM3739378","GSM3739550","GSM3740602","GSM3633854","GSM3740101","GSM3739987","GSM3740012","GSM3739464","GSM3735317","GSM3872867","GSM3735389","GSM3739356","GSM3851881","GSM3633721","GSM3739676","GSM3872794","GSM3739448","GSM3740014","GSM3633849","GSM3740652","GSM3735388","GSM3739622","GSM3735435","GSM3872843","GSM3633876","GSM3739623","GSM3872755","GSM3739615","GSM3560539","GSM3739323","GSM3739540","GSM3633842","GSM3739933","GSM3633705","GSM3740093","GSM3872751","GSM3739613","GSM3851857","GSM3739466","GSM3735415","GSM3740258","GSM3739435","GSM1241188","GSM3740558","GSM3739394","GSM3740007","GSM3740406","GSM3739988","GSM3740630","GSM3740114","GSM3739610","GSM3739339","GSM3740048","GSM3633735","GSM3872829","GSM3872831","GSM3739434","GSM3739631","GSM3740491","GSM3633741","GSM3560566","GSM3740169","GSM3739393","GSM3633754","GSM3740173","GSM3739472","GSM3739708","GSM3740172","GSM3739527","GSM3739932","GSM3739846","GSM3739905","GSM3740284","GSM3735363","GSM3739805","GSM3740763","GSM3739411","GSM3740622","GSM3851910","GSM3872913","GSM1241232","GSM3851872","GSM3739548","GSM3740032","GSM3740160","GSM3872777","GSM3739363","GSM3739587","GSM3740435","GSM3872697","GSM3735392","GSM3872738","GSM3735428","GSM3739857","GSM3633720","GSM3740199","GSM1241190","GSM3739794","GSM3739299","GSM3739811","GSM3872672","GSM3633727","GSM3851925","GSM3633752","GSM3739528","GSM3740363","GSM1241203","GSM3633893","GSM3739777","GSM3740028","GSM1241163","GSM3739374","GSM3872863","GSM3739833","GSM3739479","GSM3740369","GSM3851846","GSM3739733","GSM3739743","GSM3740350","GSM3740074","GSM3739590","GSM3633881","GSM3740229","GSM3633742","GSM3739362","GSM1241208","GSM3633763","GSM3851901","GSM3872770","GSM3872826","GSM3740748","GSM3851812","GSM3739458","GSM3851903","GSM3740487","GSM3872692","GSM3739398","GSM3739802","GSM3633831","GSM3740494","GSM3872765","GSM3740394","GSM3872678","GSM3739324","GSM3739504","GSM3739296","GSM3739825","GSM3740255","GSM3740766","GSM3851908","GSM1241248","GSM3740704","GSM3633737","GSM3560597","GSM3739728","GSM3739999","GSM3739886","GSM3735400","GSM3740092","GSM3740198","GSM3872885","GSM3740671","GSM3740168","GSM3851895","GSM3740524","GSM1241139","GSM3740683","GSM3735314","GSM3740386","GSM3560544","GSM3851928","GSM3740720","GSM3739740","GSM3739732","GSM3740453","GSM3739700","GSM3872688","GSM3740440","GSM3740718","GSM3735345","GSM3633731","GSM3739793","GSM3740567","GSM3872693","GSM3740335","GSM3872667","GSM3740523","GSM3740455","GSM3872663","GSM3739864","GSM3739436","GSM1241211","GSM3740573","GSM3739367","GSM3740204","GSM3851851","GSM3872786","GSM3740607","GSM3739957","GSM3735408","GSM3740497","GSM3633900","GSM3740508","GSM1241185","GSM3735443","GSM3633798","GSM3740546","GSM3740658","GSM3740347","GSM3740657","GSM3633718","GSM3739492","GSM3739484","GSM3633872","GSM3740080","GSM3739943","GSM3872848","GSM3740764","GSM3739626","GSM1241205","GSM3740023","GSM3633816","GSM3740513","GSM3739842","GSM3739817","GSM3633755","GSM3851823","GSM3739979","GSM3633786","GSM1241220","GSM3740097","GSM3739602","GSM3740393","GSM3851841","GSM3560551","GSM3872802","GSM3740587","GSM3740085","GSM3851937","GSM3740729","GSM3739633","GSM3872896","GSM1708825","GSM3739991","GSM3740202","GSM3740459","GSM3633905","GSM3739581","GSM3633885","GSM3633711","GSM3740418","GSM3872748","GSM3851875","GSM3560605","GSM3739391","GSM3739588","GSM3739688","GSM3740030","GSM3740353","GSM3739644","GSM3740373","GSM3735419","GSM3740768","GSM3633758","GSM3633814","GSM3735440","GSM3740355","GSM3872659","GSM3872740","GSM3739586","GSM3740075","GSM3633788","GSM3740614","GSM1241213","GSM3739639","GSM3872869","GSM3851876","GSM3740502","GSM3735368","GSM3739910","GSM3739924","GSM3735382","GSM3872739","GSM3739524","GSM3739558","GSM3633828","GSM3560527","GSM3739928","GSM3735350","GSM3739536","GSM3872707","GSM3560525","GSM3739450","GSM3739457","GSM3740103","GSM3739516","GSM3739923","GSM3872654","GSM3740669","GSM3851842","GSM3740010","GSM3872741","GSM3851883","GSM3740544","GSM3872894","GSM1241194","GSM1241214","GSM3739747","GSM3740760","GSM3851880","GSM3739381","GSM3851847","GSM3740601","GSM3739406","GSM1241224","GSM3735397","GSM3739383","GSM3740465","GSM3740714","GSM3739396","GSM3851892","GSM3739715","GSM3739526","GSM3735357","GSM3740481","GSM3872722","GSM3740349","GSM3735318","GSM3740761","GSM3872841","GSM3740215","GSM3739354","GSM3740372","GSM3633864","GSM3739721","GSM3739787","GSM3633793","GSM3739620","GSM3560630","GSM3872775","GSM3740700","GSM1241170","GSM3739338","GSM3740427","GSM3740620","GSM3851939","GSM3739614","GSM3740138","GSM3740362","GSM3560610","GSM3872711","GSM3739806","GSM3740225","GSM3872760","GSM3739724","GSM3740345","GSM3739946","GSM3739831","GSM3633865","GSM3739646","GSM3851934","GSM3872745","GSM3739465","GSM3633707","GSM3740531","GSM3633887","GSM3872887","GSM3633846","GSM3872768","GSM3740088","GSM3735366","GSM3740583","GSM3560542","GSM3739409","GSM3739852","GSM3740613","GSM3739561","GSM3851932","GSM3872656","GSM3740302","GSM3740686","GSM3739783","GSM3740120","GSM3740276","GSM3739711","GSM3739975","GSM3740273","GSM3740515","GSM3740711","GSM3633812","GSM3633724","GSM3872817","GSM3740055","GSM3739687","GSM3740530","GSM3740139","GSM3739903","GSM3739604","GSM3740245","GSM3633841","GSM3872862","GSM3872878","GSM3740570","GSM3872873","GSM3872757","GSM3740456","GSM3739902","GSM3872752","GSM3739985","GSM3740020","GSM3735379","GSM3740713","GSM3872706","GSM3851905","GSM3739713","GSM3739940","GSM3633839","GSM3739419","GSM3739926","GSM3739671","GSM3739960","GSM3739916","GSM3739444","GSM3739735","GSM3851873","GSM3739982","GSM3560628","GSM3633835","GSM3739755","GSM3739862","GSM3739878","GSM3739333","GSM3560591","GSM3739858","GSM3740400","GSM3740267","GSM3739896","GSM3633862","GSM3851868","GSM3739966","GSM3739582","GSM3739456","GSM3735394","GSM3739568","GSM3740517","GSM3739904","GSM3560554","GSM3560578","GSM3740205","GSM3739591","GSM3739954","GSM3872821","GSM3633911","GSM3739454","GSM3739439","GSM3740656","GSM3560569","GSM3740019","GSM3872728","GSM3740357","GSM3633785","GSM3633774","GSM3735365","GSM3735372","GSM3739989","GSM3740257","GSM3740555","GSM3872881","GSM3740402","GSM3740409","GSM3740646","GSM3740039","GSM3851833","GSM3735437","GSM3740447","GSM3851930","GSM3851838","GSM3740128","GSM3739847","GSM3740412","GSM3739493","GSM3633716","GSM1241199","GSM1241197","GSM3740254","GSM3739961","GSM3740038","GSM3740037","GSM3739564","GSM3739395","GSM3740726","GSM3851933","GSM3735324","GSM3739443","GSM3739770","GSM3739898","GSM3740036","GSM3740334","GSM3872716","GSM3739452","GSM3633778","GSM3872698","GSM3560590","GSM3739505","GSM3739768","GSM3560541","GSM3740041","GSM3560629","GSM3560588","GSM3560558","GSM3633886","GSM3740401","GSM3633914","GSM3872807","GSM3740710","GSM3739420","GSM3740283","GSM3851859","GSM3739804","GSM3851840","GSM3735311","GSM3633790","GSM3851839","GSM3740709","GSM3739400","GSM3740659","GSM1241162","GSM3739651","GSM3740226","GSM3739332","GSM3633903","GSM3633807","GSM3740458","GSM3739555","GSM3739322","GSM3735407","GSM3740081","GSM3872909","GSM3739761","GSM3740489","GSM3740264","GSM1241222","GSM3735399","GSM3633726","GSM3740100","GSM3739684","GSM3739675","GSM3739899","GSM3740301","GSM3851832","GSM1241175","GSM3740263","GSM1241158","GSM3740096","GSM3740025","GSM3739475","GSM3740476","GSM3739312","GSM1708826","GSM3740377","GSM3740675","GSM3872892","GSM3740697","GSM3740170","GSM3851863","GSM1241237","GSM3739488","GSM3633851","GSM3740610","GSM3740275","GSM1241226","GSM3851815","GSM3739438","GSM3740216","GSM1241215","GSM3739592","GSM3739530","GSM3740674","GSM3560615","GSM3633765","GSM3739834","GSM3740417","GSM3740762","GSM3739665","GSM3851902","GSM3735395","GSM3739781","GSM3851813","GSM3740454","GSM3872700","GSM3740248","GSM3740554","GSM3740664","GSM1241216","GSM1241233","GSM3851865","GSM3740175","GSM1241187","GSM3568641","GSM3740730","GSM3739698","GSM3739567","GSM3739792","GSM3851843","GSM3740031","GSM3740062","GSM3739428","GSM3872746","GSM3740397","GSM3740374","GSM3735359","GSM3739310","GSM3740183","GSM3740479","GSM3740562","GSM3739901","GSM3872859","GSM3740329","GSM3740154","GSM3735380","GSM3872731","GSM3739894","GSM3851827","GSM3872840","GSM3739882","GSM3739317","GSM3740441","GSM3851878","GSM3739569","GSM3739627","GSM3740001","GSM3872872","GSM3872886","GSM3872816","GSM3851915","GSM1241209","GSM1241231","GSM3735402","GSM3739346","GSM3740054","GSM3740159","GSM3872891","GSM3740189","GSM3739343","GSM3739371","GSM3740433","GSM3739495","GSM3735420","GSM3739727","GSM3851923","GSM3739522","GSM3872772","GSM3739513","GSM3739459","GSM3560581","GSM1241240","GSM3739432","GSM3872911","GSM3740518","GSM3739641","GSM3739585","GSM1241166","GSM3739950","GSM3633853","GSM3851817","GSM3739399","GSM3739837","GSM3740232","GSM3740577","GSM3739778","GSM3735386","GSM3739731","GSM3739870","GSM3872721","GSM3740188","GSM3633753","GSM3872689","GSM3739355","GSM3740681","GSM3739365","GSM3739598","GSM3633892","GSM3633771","GSM3739616","GSM3872717","GSM3740540","GSM3851936","GSM3735321","GSM3735319","GSM3740662","GSM3740328","GSM3740638","GSM3872704","GSM3560526","GSM3560617","GSM3740117","GSM3740520","GSM3851811","GSM3739606","GSM3740270","GSM3735374","GSM3739451","GSM3740410","GSM3633794","GSM3740163","GSM3560624","GSM3740109","GSM1708817","GSM3872820","GSM3740691","GSM3735431","GSM3740543","GSM3740087","GSM3740667","GSM1241196","GSM3739521","GSM3739369","GSM3740444","GSM3740579","GSM3739955","GSM3740478","GSM3735391","GSM3633748","GSM1241191","GSM3739473","GSM1241161","GSM3851913","GSM3872660","GSM3740596","GSM3633797","GSM3872780","GSM3633732","GSM3740155","GSM3740450","GSM3735434","GSM3872858","GSM3739300","GSM3872864","GSM1241221","GSM1241195","GSM3739643","GSM3739883","GSM3740181","GSM3740179","GSM3739929","GSM3740424","GSM3739789","GSM3740595","GSM3740680","GSM3739996","GSM3633891","GSM3740715","GSM3872726","GSM3739532","GSM3739314","GSM3740472","GSM3739776","GSM3735358","GSM3872876","GSM3633777","GSM3740035","GSM3740082","GSM3739784","GSM3740240","GSM3872861","GSM3739939","GSM1708820","GSM3739876","GSM3872724","GSM3851836","GSM3739417","GSM3740468","GSM3560613","GSM3739416","GSM3633706","GSM3739814","GSM3740733","GSM3740712","GSM3633734","GSM3739922","GSM1241234","GSM3851819","GSM3739692","GSM3851870","GSM3872830","GSM3740009","GSM3739891","GSM3739840","GSM3740178","GSM3739552","GSM3740143","GSM3872734","GSM3739993","GSM3633904","GSM3739441","GSM3740194","GSM3739388","GSM3739594","GSM3739353","GSM3740572","GSM3851816","GSM3739549","GSM3739308","GSM3633815","GSM3633919","GSM3739912","GSM3739888","GSM3739695","GSM3739753","GSM3739556","GSM3739608","GSM3851850","GSM3735416","GSM3851845","GSM3560622","GSM3740171","GSM3735322","GSM3740322","GSM3739709","GSM3872871","GSM3735326","GSM3740702","GSM3872842","GSM3739701","GSM3739571","GSM3633918","GSM3740070","GSM1241189","GSM3739517","GSM3633907","GSM3872805",]),
        #expand("out/hs/seq/gsms/{gsm}.tsv", gsm=["GSM2666941", "GSM2666942", "GSM2666943"]),
        #expand("out/mm/seq/gses/{gse}.tsv", gse=["GSE89477", "GSE89263"])
        #"out/mm/seq/postquant_filter/gsm_stats.tsv"
        # expand("flags/{organism}/{platform}/pca/{n_genes}_{scale}/flag",
        #         organism=['mm'],
        #         platform=['chip'],
        #         n_genes=config['pca_n_genes'],
        #         scale=config['pca_scale'])
        # expand("out/{organism}/{platform}/pca_fgsea/"
        #        "{n_genes}_{scale}_{max_comp}_{var_threshold}_{gsea_param}/"
        #        "prepared/{geneset_name}.tsv",
        #     organism='mm',
        #     platform=['chip', "seq"],
        #     n_genes="6000",
        #     scale="linear",
        #     max_comp="10",
        #     var_threshold="0.02",
        #     gsea_param=["0","1"],
        #     geneset_name=["AGE_CD8_TCELLS"])#['HALLMARK_HYPOXIA', "GSE120744_TREM_SIGNATURE" ])#'HALLMARK_PANCREAS_BETA_CELLS'
            #               'HALLMARK_PI3K_AKT_MTOR_SIGNALING', 'HALLMARK_SPERMATOGENESIS',
            #               'HALLMARK_FATTY_ACID_METABOLISM', 'HALLMARK_BILE_ACID_METABOLISM',
            #               'HALLMARK_P53_PATHWAY', 'HALLMARK_MYOGENESIS', 'HALLMARK_PROTEIN_SECRETION',
            #               'HALLMARK_UV_RESPONSE_DN', 'HALLMARK_ANGIOGENESIS', 'HALLMARK_NOTCH_SIGNALING',
            #               'HALLMARK_MYC_TARGETS_V2', 'HALLMARK_TNFA_SIGNALING_VIA_NFKB', 'HALLMARK_KRAS_SIGNALING_DN',
            #               'HALLMARK_HEDGEHOG_SIGNALING', 'HALLMARK_APICAL_SURFACE', 'HALLMARK_MYC_TARGETS_V1',
            #               'HALLMARK_ALLOGRAFT_REJECTION', 'HALLMARK_CHOLESTEROL_HOMEOSTASIS',
            #               'HALLMARK_ANDROGEN_RESPONSE', 'HALLMARK_E2F_TARGETS', 'HALLMARK_GLYCOLYSIS',
            #               'HALLMARK_DNA_REPAIR', 'HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION',
            #               'HALLMARK_IL6_JAK_STAT3_SIGNALING', 'HALLMARK_OXIDATIVE_PHOSPHORYLATION',
            #               'HALLMARK_UNFOLDED_PROTEIN_RESPONSE', 'HALLMARK_REACTIVE_OXYGEN_SPECIES_PATHWAY',
            #               'HALLMARK_INFLAMMATORY_RESPONSE', 'HALLMARK_UV_RESPONSE_UP',
            #               'HALLMARK_WNT_BETA_CATENIN_SIGNALING', 'HALLMARK_INTERFERON_ALPHA_RESPONSE',
            #               'HALLMARK_G2M_CHECKPOINT', 'HALLMARK_IL2_STAT5_SIGNALING', 'HALLMARK_APOPTOSIS',
            #               'HALLMARK_INTERFERON_GAMMA_RESPONSE', 'HALLMARK_ESTROGEN_RESPONSE_LATE',
            #               'HALLMARK_COAGULATION', 'HALLMARK_XENOBIOTIC_METABOLISM', 'HALLMARK_COMPLEMENT',
            #               'HALLMARK_ADIPOGENESIS', 'HALLMARK_TGF_BETA_SIGNALING', 'HALLMARK_MITOTIC_SPINDLE',
            #               'HALLMARK_MTORC1_SIGNALING', 'HALLMARK_APICAL_JUNCTION', 'HALLMARK_KRAS_SIGNALING_UP',
            #               'HALLMARK_PEROXISOME', 'HALLMARK_ESTROGEN_RESPONSE_EARLY', 'HALLMARK_HEME_METABOLISM'])

#############################################FUNCTIONS##################################################################


def get_filtered_gse_gsm_map(gsm_df_file, organism,
                             min_gsm=None, max_gsm=None,
                             gpl_df_file=None, # if to be filtered by supported gpl
                             srr_df_file=None, min_spots=None, max_spots=None, # if to be filtered by the number of spots
                             gsm_qc_file=None, min_exp_genes=None,
                             n_batches=None, batch_n=None):
    """
    Takes FILES not links. Links to checkpoints are called inside other functions.
    Abstracts GSM GSE filtering on every level of the pipeline. Called in other functions to obtain proper mappings and
    filtered lists of GSM/GSE
    """
    global glob_gsm_srr_map
    global glob_gse_gsm_map

    print("Filtering GSM by metadata...")
    organism_gsm = {"hs": "Homo sapiens", "mm": "Mus musculus", "rn": "Rattus norvegicus"}[organism]
    gsm_df = pd.read_csv(gsm_df_file, sep='\t')
    gsm_df = gsm_df[gsm_df["ORGANISM"] == organism_gsm]
    gsm_df = gsm_df[gsm_df["LIBRARY_SELECTION"] == "cDNA"]
    gsm_df = gsm_df[gsm_df["LIBRARY_STRATEGY"] == "RNA-Seq"]

    if gpl_df_file:
        print("Filtering GSM by platform...")
        gpl_df = pd.read_csv(gpl_df_file, sep='\t')
        gpl_df = gpl_df[gpl_df["LAB"]==organism.upper()]
        gpl_list = gpl_df["GPL"].tolist()
        gsm_df = gsm_df[gsm_df["GPL"].isin(gpl_list)]

    gsm_df = gsm_df[["GSE", "GSM"]]
    gsm_df = gsm_df.drop_duplicates()

    if min_spots:
        print("Filtering GSM by number of spots...")
        if organism not in glob_gsm_srr_map:
            print("Saving srr df to global var")
            srr_df = pd.read_csv(srr_df_file, sep="\t")
            gsm_list = list(set(gsm_df['GSM'].tolist()))
            srr_df = srr_df[srr_df["GSM"].isin(gsm_list)]
            glob_gsm_srr_map[organism] = srr_df

        print("Retrieving GSM SRR map from global...")
        srr_df = glob_gsm_srr_map[organism]
        srr_df = srr_df[['GSM', 'SPOTS']]
        srr_df = srr_df.groupby(['GSM']).sum().reset_index()
        srr_df = srr_df.astype({'GSM': 'str', 'SPOTS':'float'})
        srr_df = srr_df[(srr_df['SPOTS'] >= min_spots) & (srr_df['SPOTS'] <= max_spots)]
        filtered_gsm_list = srr_df['GSM'].tolist()
        gsm_df = gsm_df[gsm_df['GSM'].isin(filtered_gsm_list)]

    if gsm_qc_file:
        print("Filtering GSM by number of expressed genes...")
        gsm_qc_df = pd.read_csv(gsm_qc_file, sep='\t')
        gsm_qc_df = gsm_qc_df[gsm_qc_df["N_EXP_GENES"] >= min_exp_genes]
        filtered_gsm_list = gsm_qc_df["GSM"].tolist()
        gsm_df = gsm_df[gsm_df['GSM'].isin(filtered_gsm_list)]

    if min_gsm:
        print("Filtering by number of passing GSM per GSE...")
        gse_counts_df = gsm_df['GSE'].value_counts().reset_index()
        gse_counts_df = gse_counts_df.rename(columns={"index":"GSE", "GSE":"COUNT"})
        gse_counts_df = gse_counts_df[(gse_counts_df['COUNT'] >= min_gsm) & (gse_counts_df['COUNT'] <= max_gsm)]
        passing_gse = gse_counts_df['GSE'].tolist()
        gsm_df = gsm_df[gsm_df['GSE'].isin(passing_gse)]
        print("Further sub sampling for SRR df...")
        gsm_list = list(set(gsm_df['GSM'].tolist()))
        glob_gsm_srr_map[organism] = glob_gsm_srr_map[organism][glob_gsm_srr_map[organism]["GSM"].isin(gsm_list)]

    print("Done.")
    return gsm_df

#############################################RULES######################################################################

rule sm_download:
    '''
    Takes search output .txt, parses out all links and downloads them. (wget -nc) Existing up to date files out dir
    doesn't reload. Writes completion flag in the end.
    '''
    resources:
        time=60*24*2,
        download_res=1,
        writing_res=1,
        priority=5
    input:
        "input/{organism}/{platform}/gds_search_result.txt"
    output:
        "flags/{organism}/{platform}/sm_download.flag"
    params:
        sm_download_dir="out/{organism}/{platform}/series_matrices"
    message: "Downloading series matrices for {wildcards.organism} {wildcards.platform}..."
    log: "logs/{organism}/{platform}/sm_download.log"
    shell:
        "scripts/bash/download_sm.sh {input} {params.sm_download_dir} "
        " > {log} 2>&1 &&"
        " touch {output}"

checkpoint extract_sm_metadata:
    input:
        rules.sm_download.output
    output:
        gsm_df="out/{organism}/{platform}/sm_metadata/gsm.tsv",
        gse_df="out/{organism}/{platform}/sm_metadata/gse.tsv"
    params:
        sm_download_dir=rules.sm_download.params.sm_download_dir
    message: "Aggregating metadata from series matrices {wildcards.organism} {wildcards.platform} ..."
    log: "logs/{organism}/{platform}/sm_metadata.log"
    shell:
        "python scripts/python/parse_sm_metadata.py {params.sm_download_dir} {output.gse_df} {output.gsm_df}"
        " > {log} 2>&1"

#############################################RNASEQ#####################################################################
rule sra_accession_df_download:
    resources:
        download_res=1,
        writing_res=1,
        mem_ram=2
    output:
        temp("out/data/sra_accession_raw.tab")
    message: "Downloading SRA accession table..."
    log: "logs/sra_accession_df_download.log"
    shell:
        "wget -O {output} {config[sra_accession_df]}"
        " > {log} 2>&1"

rule get_srr_df:
    resources:
        writing_res=1
    input:
        rules.sra_accession_df_download.output
    output:
        "out/data/srr_gsm_spots.tsv"
    message: "Cleaning SRR df..."
    log: "logs/get_srr_df.log"
    conda: "envs/r_scripts.yaml"
    shell:
        "scripts/bash/clean_sra_accession_df.sh {input} {output}"
        " > {log} 2>&1 &&"
        " Rscript scripts/R/clean_sra_accession_df.R {output}"
        " >> {log} 2>&1"

rule sra_download:
    resources:
        download_res=1,
        writing_res=1,
        mem_ram=2
    priority: 3
    output: temp("out/{organism}/seq/sra/{srr}.sra")
    log:    "logs/{organism}/seq/sra_download/{srr}.log"
    message: "Downloading {wildcards.srr} ({wildcards.organism})..."
    shadow: "shallow"
    conda: "envs/quantify.yaml"
    shell:
        "scripts/bash/download_sra.sh {wildcards.srr} {output}"
        " > {log} 2>&1"

rule sra_fastqdump:
    resources:
        writing_res=1
    input:
        rules.sra_download.output
    output:
        fastq_dir=temp(directory("out/{organism}/seq/fastq/{srr}")),
    log:    "logs/{organism}/seq/sra_fastqdump/{srr}.log"
    message: "fastq-dump {wildcards.srr} ({wildcards.organism})"
    conda: "envs/quantify.yaml"
    shadow: "shallow"
    shell:
        "fastq-dump --outdir {output.fastq_dir} --split-3 {input}"
        " > {log} 2>&1"

rule fastq_kallisto:
    resources:
        mem_ram=lambda wildcards: config["quant_mem_ram"][wildcards.organism]
    priority: 2
    input:
        fastq_dir=rules.sra_fastqdump.output,
        refseq="input/{organism}/seq/refseq_kallisto"
    output:
        # before uncommenting h5 remove files of older versions as described in README. See NOTES on h5
        #h5=protected("out/{organism}/seq/kallisto/{srr}/abundance.h5"),
        tsv=protected("out/{organism}/seq/kallisto/{srr}/abundance.tsv"),
        json=protected("out/{organism}/seq/kallisto/{srr}/run_info.json")
    log: "logs/{organism}/seq/fastq_kallisto/{srr}.log"
    message: "Kallisto: {wildcards.srr} ({wildcards.organism})"
    conda: "envs/quantify.yaml"
    shadow: "shallow"
    shell:
        "scripts/bash/quantify.sh {wildcards.srr} {input.fastq_dir} {input.refseq}"
        " out/{wildcards.organism}/seq/kallisto/{wildcards.srr}"# output dir 
        " {config[n_bootstrap]} " 
        " > {log} 2>&1"

def get_srr_files_for_gsm(wildcards):
    print(f"Getting SRR for {wildcards.gsm}")
    global glob_gsm_srr_map

    # for speed. Otherwise it takes forever
    if wildcards.organism not in glob_gsm_srr_map:
        print("Saving srr df to global var")
        srr_df_file = "out/data/srr_gsm_spots.tsv"
        srr_df = pd.read_csv(srr_df_file, sep="\t")
        gsm_df = get_filtered_gse_gsm_map(f"out/{wildcards.organism}/seq/sm_metadata/gsm.tsv", wildcards.organism)
        gsm_list = list(set(gsm_df['GSM'].tolist()))
        srr_df = srr_df[srr_df["GSM"].isin(gsm_list)]
        glob_gsm_srr_map[wildcards.organism] = srr_df

    srr_df = glob_gsm_srr_map[wildcards.organism]
    srr_list = srr_df[srr_df['GSM']==wildcards.gsm]["SRR"].tolist()
    srr_list = list(set(srr_list)) # removing possible duplicates
    srr_files = \
        expand(rules.fastq_kallisto.output.tsv,
        srr=srr_list,
        organism=wildcards.organism)

    return srr_files

rule srr_to_gsm:
    resources:
        mem_ram=1
    input:
        srr_files=ancient(get_srr_files_for_gsm),
        transcript_gene=ancient("input/{organism}/seq/transcript_gene.tsv")
    output:
        gsm_file="out/{organism}/seq/gsms/{gsm}.tsv",
        gsm_qc="out/{organism}/seq/gsms_qc/{gsm}_qc.tsv"
    log: "logs/{organism}/seq/srr_to_gsm/{gsm}.log"
    message: "Aggregating {wildcards.gsm} ({wildcards.organism})..."
    conda: "envs/r_scripts.yaml"
    shell:
        "Rscript scripts/R/srr_to_gsm.R {output} {input.transcript_gene} {input.srr_files}"
        " > {log} 2>&1"

def get_filtered_gsm_files_prequant(wildcards, **kwargs):
    print("Getting GSM files for quantification...")
    gse_gsm_map = get_filtered_gse_gsm_map(
        gsm_df_file=checkpoints.extract_sm_metadata.get(**wildcards).output.gsm_df,
        organism=wildcards.organism,
        **kwargs
    )
    gsm_list = list(set(gse_gsm_map['GSM'].tolist()))

    print("Done.")
    return \
        expand("out/{organism}/seq/gsms/{gsm}.tsv",
            organism=wildcards.organism, gsm=gsm_list)

checkpoint aggregate_gsm_qc_df:
    input:
        lambda wildcards:
        ancient(get_filtered_gsm_files_prequant(wildcards,
            min_gsm=2, max_gsm=200,
            gpl_df_file="input/gpl.tsv", # if to be filtered by supported gpl
            srr_df_file=str(rules.get_srr_df.output),
            # if to be filtered by the number of spots
            min_spots=int(config["min_spots"][wildcards.organism]),
            max_spots=int(config["max_spots"][wildcards.organism])))
    output:
        gsm_qc_df="out/{organism}/seq/gsm_qc.tsv"
    run:
        print("Aggregating GSM qc df...")
        n_exp_genes=\
            [
                sum(pd.read_csv(gsm_file,sep='\t')['EST_COUNTS']!=0)
                for gsm_file in input
            ]
        gsm_id_list=\
            [
                re.findall("GSM\d+", gsm_file)[0]
                for gsm_file in input
            ]
        gsm_qc_df = pd.DataFrame(
            {'GSM': gsm_id_list,
             'N_EXP_GENES': n_exp_genes
             })
        gsm_qc_df.to_csv(str(output),sep='\t',index=False)
        print("Done.")

def get_gsm_files_for_gse(wildcards):
    global glob_postquant_gsm_gse_map
    print(f"Getting GSM files for {wildcards.gse}...")
    if wildcards.organism not in glob_postquant_gse_gsm_map:
        glob_postquant_gse_gsm_map[wildcards.organism] = \
            get_filtered_gse_gsm_map(
                gsm_df_file=checkpoints.extract_sm_metadata.get(**wildcards).output.gsm_df,
                min_gsm=2, max_gsm=200,
                gpl_df_file="input/gpl.tsv", # if to be filtered by supported gpl
                srr_df_file=str(rules.get_srr_df.output),
                # if to be filtered by the number of spots
                min_spots=int(config["min_spots"][wildcards.organism]),
                max_spots=int(config["max_spots"][wildcards.organism]),
                gsm_qc_file=str(checkpoints.aggregate_gsm_qc_df.get(**wildcards).output),
                min_exp_genes=int(config["min_exp_genes"]),
                organism=wildcards.organism
        )
    gse_gsm_map = glob_postquant_gsm_gse_map[wildcards.organism]
    gsm_list = gse_gsm_map[gse_gsm_map['GSE']==wildcards.gse].tolist()

    return \
        expand(rules.srr_to_gsm.output.gsm_file,
            organism=wildcards.organism, gsm=gsm_list)

rule gsm_to_gse_cpm:
    '''
    Aggregates GSM to sorted by TPM values, CPM GSE
    '''
    input:
        gsm_files=ancient(get_gsm_files_for_gse),
        gene_mapping=ancient("input/{organism}/seq/ensembl_symbol_entrez.tsv")
    output:
        gse=protected("out/{organism}/seq/exp_mat/{gse}.tsv")
    log: "logs/{organism}/gsm_to_gse_cpm/{gse}.log"
    message: "Aggregating {wildcards.gse} ({wildcards.organism})..."
    shadow: "shallow"
    conda: "envs/r_scripts.yaml"
    shell:
        "Rscript scripts/R/gsm_to_gse.R {input.gene_mapping} {output.gse} {input.gsm_files}"
        " > {log} 2>&1"

rule gsm_to_gse_counts:
    '''
    Aggregates unsorted raw counts GSE
    '''
    input:
        gsm_files=ancient(get_gsm_files_for_gse),
        gene_mapping=ancient("input/{organism}/seq/ensembl_symbol_entrez.tsv")
    output:
        gse=protected("out/{organism}/seq/gses_counts/{gse}.tsv")
    log: "logs/{organism}/gsm_to_gse_counts/{gse}.log"
    message: "Aggregating {wildcards.gse} ({wildcards.organism})..."
    shadow: "shallow"
    conda: "envs/r_scripts.yaml"
    shell:
        "Rscript scripts/R/gsm_to_gse_raw_counts.R {input.gene_mapping} {output.gse} {input.gsm_files}"
        " > {log} 2>&1"

checkpoint aggregate_gse_qc_df:
    """
    DOESN'T PULL ANY OTHER RULES YET. Just aggregates info on existing files in folder
    Aggregates GSE qc df with GSE ids and number of GSM per GSE
    """
    # input:
    #     lambda wildcards:
    #         get_filtered_gse_files(wildcards,
    #             min_gsm=2, max_gsm=200,
    #             gpl_df_file="input/gpl.tsv", # if to be filtered by supported gpl
    #             srr_df_file=str(rules.get_srr_df.output),
    #             # if to be filtered by the number of spots
    #             min_spots=int(config["min_spots"][wildcards.organism]),
    #             max_spots=int(config["max_spots"][wildcards.organism]))
    output:
        "out/{organism}/seq/gse_cpm_qc.tsv"
    run:
        print(f"Aggregating GSE qc df for {wildcards.organism}...")
        gses_dir=f"out/{wildcards.organism}/seq/exp_mat"
        gse_files = [gses_dir + "/" + f for f in listdir(gses_dir)]
        n_gsm_list=\
            [
                len(pd.read_csv(gse_file,sep='\t').columns)
                for gse_file in gse_files
            ]
        gse_id_list=\
            [
                re.findall("GSE\d+", gse_file)[0]
                for gse_file in gse_files
            ]
        gse_qc_df = pd.DataFrame(
            {'GSE': gse_id_list,
             'N_GSM': n_gsm_list
             })

        gse_qc_df.to_csv(str(output),sep='\t',index=False)
        print("Done.")


#############################################CHIP_ROOT##################################################################
rule extract_exp_mat:
    '''
    Extracts expression matrices from files and writes QC report for downstream filtering.
    gq_df: TAG	N_GSM	GPL	HAS_EXP_MAT	LOGAV	LINMAX	LOGMAX	N_GENES	HAS_NEGATIVE_VALUES	PROCESSED
    Writes empty exp table if error produced.
    '''
    resources:
        time=40
    input:
        gpl_dir="input/{organism}/chip/platform_annotation"
    params:
        sm="out/{organism}/chip/series_matrices/{tag}_series_matrix.txt.gz"
    output:
        exp_table=protected("out/{organism}/chip/exp_mat/{tag}.tsv"),
        qc_report=protected("out/{organism}/chip/exp_mat/{tag}_qc.tsv")
    log: "logs/{organism}/chip/extract_exp_mat/{tag}.log"
    message: "Extracting expression table from {wildcards.tag}_series_matrix.txt.gz ({wildcards.organism})..."
    conda: "envs/r_scripts.yaml"
    shell:
        "Rscript scripts/R/get_exp_table.R {params.sm} {input.gpl_dir} {output.exp_table} {output.qc_report}"
        " > {log} 2>&1"

def get_filtered_sm_qc_files(wildcards):
    print("Filtering SM files...")
    gse_df_file = checkpoints.extract_sm_metadata.get(platform="chip", organism=wildcards.organism).output.gse_df
    print(gse_df_file)
    gse_df = pd.read_csv(gse_df_file, sep="\t")
    print(gse_df)
    gse_df = gse_df[(gse_df['NUMBER_GSM']>=config['min_gsm']) & (gse_df['NUMBER_GSM']<=config['max_gsm'])]
    print("Filtering SM by list of supported GPL...")
    gpl_dir = f"input/{wildcards.organism}/chip/platform_annotation"
    gpl_list = [f for f in listdir(gpl_dir) if isfile(join(gpl_dir, f))]
    gpl_list = [re.search('(.+?).3col.gz', gpl).group(1) for gpl in gpl_list]
    gse_df = gse_df[gse_df["GPL"].isin(gpl_list)]
    print("Removing super series...")
    gse_df = gse_df[gse_df["SUPER_SERIES_OF"].isna()]

    filtered_sm_tags=gse_df["GSE"].to_list()

    filtered_sm_qc_files = \
        expand('out/{organism}/chip/exp_mat/{tag}_qc.tsv',
        organism=wildcards.organism,
        tag=filtered_sm_tags)
    print("Done.")
    return filtered_sm_qc_files

rule get_exp_mat_qc_df:
    input: get_filtered_sm_qc_files
    output: "out/{organism}/{platform}/exp_mat_qc.tsv"
    run:
        print(f"Aggregating QC df for SM {wildcards.organism}...")
        qc_df_list=[pd.read_csv(qc_file,sep='\t') for qc_file in input]
        qc_df=pd.concat(qc_df_list)
        qc_df.to_csv(str(output),sep='\t',index=False)
        print("Done.")

#############################################INPUT_FUNCTION_QC##########################################################


#############################################PCA########################################################################
# def get_exp_mat_file(wildcards):
#     print(wildcards.platform)
#     if wildcards.platform=="chip":
#         return f"out/{wildcards.organism}/chip/exp_mat/{wildcards.tag}.tsv"
#     if wildcards.platform=="seq":
#         return f"out/{wildcards.organism}/seq/exp_mat/{wildcards.tag}.tsv"

rule pca:
    '''
    Takes exp_mat as an input and outputs first 
    1) 10 (or less) PC components loadings scores
    2) Stats- explained var, etc. 
    '''
    resources:
        time = 20
    input:
        # get_exp_mat_file
        "out/{organism}/{platform}/exp_mat/{tag}.tsv"
    output:
        protected("out/{organism}/{platform}/pca/{n_genes}_{scale}/{tag}.rds"),
    message:
        "Performing PCA on {wildcards.tag} \n"
        " Organism: {wildcards.organism} \n"
        " Platform: {wildcards.platform} \n"
        " Number of genes considered: {wildcards.n_genes} \n"
        " Scale: {wildcards.scale}"
    log: "logs/{organism}/{platform}/pca/{n_genes}_{scale}/{tag}.log"
    conda: "envs/r_scripts.yaml"
    shell:
        "Rscript scripts/R/pca.R {input} {output} {wildcards.n_genes} {wildcards.scale}"
        " > {log} 2>&1"


def get_filtered_exp_mat_files(wildcards, min_gsm=int, max_gsm=int, n_genes=int,
                               logav_min=int(config["chip_logav_min"]), logav_max=int(config["chip_logav_max"]),
                               logmax_max=int(config["chip_logmax_max"]), linmax_max=int(config["chip_linmax_max"]),
                               allow_negative_val=False):
    """
    QC logic and rooting for choosing files for downstream analysis.
    allow_negative_val-flag tells if negative values allowed in exp mat, which is the case for some chips.
    """
    print(f"Getting filtered exp matrices for {wildcards.organism} {wildcards.platform}...")
    if wildcards.platform=="chip":
        sm_qc_df_file = f"out/{wildcards.organism}/chip/exp_mat_qc.tsv"
        sm_qc_df = pd.read_csv(sm_qc_df_file, sep="\t")
        sm_qc_df = sm_qc_df.astype({'HAS_NEGATIVE_VALUES': 'bool'})
        sm_qc_df = sm_qc_df[sm_qc_df['PROCESSED']==True]
        sm_qc_df = sm_qc_df[(sm_qc_df['N_GSM']>=min_gsm) & (sm_qc_df['N_GSM']<=max_gsm)]
        sm_qc_df = sm_qc_df[(sm_qc_df['LOGAV']>=logav_min) & (sm_qc_df['LOGAV']<=logav_max)]
        sm_qc_df = sm_qc_df[sm_qc_df['LINMAX']<=linmax_max]
        sm_qc_df = sm_qc_df[sm_qc_df['LOGMAX']<=logmax_max]
        # TODO: for some reason pca.R does not handles zero variance filtering in this dataset. Make it work later.
        sm_qc_df = sm_qc_df[sm_qc_df['TAG']!="GSE75171"]
        sm_qc_df = sm_qc_df[(sm_qc_df['HAS_NEGATIVE_VALUES']==False) | allow_negative_val]
        sm_qc_df = sm_qc_df[sm_qc_df['N_GENES']>=n_genes]
        exp_mat_tags = sm_qc_df["TAG"].tolist()

    elif wildcards.platform=="seq":
        gse_qc_df_file = str(checkpoints.aggregate_gse_qc_df.get(**wildcards).output)
        gse_qc_df = pd.read_csv(gse_qc_df_file, sep="\t")
        gse_qc_df = gse_qc_df[(gse_qc_df['N_GSM'] >= min_gsm) & (gse_qc_df['N_GSM'] <= max_gsm)]
        exp_mat_tags = gse_qc_df["GSE"].tolist()
        # exp_mat_files = \
        #     expand(
        #         rules.gsm_to_gse_cpm.output.gse,
        #         organism=wildcards.organism,
        #         gse=exp_mat_tags
        #     )
    return exp_mat_tags

rule get_pc_list_adapter:
    '''
    Apparently it is impossible to pass 4k arguments with bash. This is workaround with concatenation of arguments to a 
    single list. Note: it is not a problem with large number of inputs if they passed to 'run' section.
    '''
    input:
        ancient(
            lambda wildcards:
            expand(rules.pca.output,
                organism=wildcards.organism,
                platform=wildcards.platform,
                n_genes=wildcards.n_genes,
                scale=wildcards.scale,
                tag=get_filtered_exp_mat_files(
                    wildcards,
                    min_gsm=int(config["pca_min_gsm"]),
                    max_gsm=int(config["pca_max_gsm"]),
                    n_genes=int(wildcards.n_genes))))
    output:
        "out/{organism}/{platform}/pca/{n_genes}_{scale}.list"
    run:
        with open(str(output), 'w') as f:
            for file in input:
                f.write("%s\n" % file)

rule get_pc_list:
    '''
    Aggregates pca components to single list of ranked lists. Filters components by QC.
    QC params:
    1) Explained valiance threshold. 2% for starters.
    2) No more then 10pc per df.
    If results look good conditions might be relaxed.
    '''
    resources:
        mem_ram=16
    input:
        rules.get_pc_list_adapter.output
    message:
        "Getting PC list {wildcards.organism} {wildcards.platform} \n"
        " Number of genes considered: {wildcards.n_genes} \n"
        " Scale of original dataset: {wildcards.scale} \n"
        " Explained variance % threshold: {wildcards.var_threshold} \n"
        " Max PC components for 1 dataset: {wildcards.max_comp} \n"
    log: "logs/{organism}/{platform}/get_pc_list/{n_genes}_{scale}_{max_comp}_{var_threshold}.log"
    output:
        "out/{organism}/{platform}/pca/{n_genes}_{scale}_{max_comp}_{var_threshold}_PCList.rds"
    conda: "envs/r_scripts.yaml"
    shell:
        "Rscript scripts/R/get_pc_list.R {output} {wildcards.max_comp} {wildcards.var_threshold} {input}"
        " > {log} 2>&1"

# TODO: remove first two lines in genesets inside the script. Artifact from GQ
# rule fgsea_genesets:
#     '''
#     Performs fgsea against list of PC components. Outputs ranked list of results with NES.
#     '''
#     resources:
#         mem_ram=32,
#         threads=8
#     input:
#         pc_list=rules.get_pc_list.output,
#         geneset="input/{organism}/genesets/{geneset_name}",
#     output:
#         "out/{organism}/{platform}/pca_fgsea/"
#         "{n_genes}_{scale}_{max_comp}_{var_threshold}_{gsea_param}/"
#         "raw/{geneset_name}.tsv"
#     message:
#         "Performing GSEA {wildcards.organism} seq \n"
#         " Geneset: {wildcards.geneset_name} \n"
#         " Number of genes considered: {wildcards.n_genes} \n"
#         " Scale of original dataset: {wildcards.scale} \n"
#         " Explained variance threshold %: {wildcards.var_threshold} \n"
#         " Max PC components for 1 dataset: {wildcards.max_comp} \n"
#         " FGSEA weight parameter: {wildcards.gsea_param}"
#     log:
#         "logs/{organism}/{platform}/fgsea_genesets/"
#         "{n_genes}_{scale}_{max_comp}_{var_threshold}_{gsea_param}/"
#         "{geneset_name}.log"
#     conda: "envs/fgsea.yaml"
#     shell:
#         "Rscript scripts/R/fgsea_geneset.R {input.pc_list} {input.geneset} {output} {wildcards.gsea_param}"
#         " > {log} 2>&1"
#
# rule prepare_pca_fgsea_result:
#     input:
#         gsea_results=rules.fgsea_genesets.output,
#         gse_df="out/{organism}/{platform}/sm_metadata/gse.tsv"
#     output:
#         "out/{organism}/{platform}/pca_fgsea/{n_genes}_{scale}_{max_comp}_{var_threshold}_{gsea_param}/"
#         "prepared/{geneset_name}.tsv"
#     message:
#         "Preparing results for PCA query. {wildcards.organism} seq \n"
#         " Geneset: {wildcards.geneset_name} \n"
#         " Number of genes considered: {wildcards.n_genes} \n"
#         " Scale of original dataset: {wildcards.scale} \n"
#         " Explained variance threshold %: {wildcards.var_threshold} \n"
#         " Max PC components for 1 dataset: {wildcards.max_comp} \n"
#         " FGSEA weight parameter: {wildcards.gsea_param}"
#     log:
#         "logs/{organism}/{platform}/prepare_pca_fgsea_result/"
#         "{n_genes}_{scale}_{max_comp}_{var_threshold}_{gsea_param}/"
#         "{geneset_name}.log"
#     conda: "envs/r_scripts.yaml"
#     shell:
#         "Rscript scripts/R/pca_prepare_results.R {input.gsea_results} {input.gse_df} {output}"
#         " > {log} 2>&1"


#########################################WGCNA##########################################################################
# # TODO: CHANGE GSES in seq to exp_mat so that it is assessable from downstream rules with similar path pattern
# rule wgcna:
#     input:
#         "out/{organism}/{platform}/exp_mat/{tag}.tsv"
#     output:
#         modules="out/{organism}/{platform}/wgcna/{n_genes}_{cor_type}_{network}_{scale}/{tag}/modules.tsv",
#         eigengenes="out/{organism}/{platform}/wgcna/{n_genes}_{cor_type}_{network}_{scale}/{tag}/eigengenes.tsv",
#         stats="out/{organism}/{platform}/wgcna/{n_genes}_{cor_type}_{network}_{scale}/{tag}/stats.txt"
#     log: "logs/{organism}/{platform}/wgcna/{n_genes}_{cor_type}_{network}_{scale}/{tag}.log"
#     message: "Performing WGCNA on {wildcards.tag} ({wildcards.organism}, {wildcards.platform})..."
#     conda: "envs/r_scripts.yaml"
#     shell:
#         "Rscript scripts/R/wgcna.R {input} {output.modules} {output.eigengenes} {output.stats} "
#         " {wildcards.n_genes} {wildcards.cor_type} {wildcards.network} {wildcards.scale}"
#         " > {log} 2>&1"
# #
# checkpoint get_wgcna_stats:
#     input:
#         lambda wildcards:
#             expand("out/{organism}/{platform}/wgcna/{n_genes}_{cor_type}_{network}_{scale}/{tag}/stats.txt",
#                 organism=wildcards.organism,
#                 platform=wildcards.platform,
#                 n_genes=wildcards.n_genes,
#                 cor_type=wildcards.cor_type,
#                 network=wildcards.network,
#                 scale=wildcards.scale,
#                 tag=get_filtered_exp_mat_files(wildcards, int(config["min_gsm_wgcna"]),
#                 int(config["max_gsm_wgcna"]), int(wildcards.n_genes)))
#     output:
#         "out/{organism}/{platform}/wgcna/{n_genes}_{cor_type}_{network}_{scale}_stats.tsv"
#     conda: "envs/r_scripts.yaml"
#     log: "logs/{organism}/{platform}/get_wgcna_stats/{n_genes}_{cor_type}_{network}_{scale}.log"
#     message:
#         "Getting WGCNA stats. \n"
#         " Organism: {wildcards.organism} \n"
#         " Platform: {wildcards.platform} \n"
#         " Number of genes considered: {wildcards.n_genes} \n"
#         " Correlation metric: {wildcards.cor_type} \n"
#         " Network type: {wildcards.network} \n"
#         " Scale: {wildcards.scale} \n"
#     run:
#         qc_df_list=[pd.read_csv(qc_file,sep='\t') for qc_file in input]
#         qc_df=pd.concat(qc_df_list)
#         qc_df.to_csv(str(output),sep='\t',index=False)
#         print("Done.")


##############################################KS_RESULTS#################################

rule ks_genesets:
    '''
    Performs ks test on list of PC components. Output ranked by p-val
    '''
    resources:
        mem_ram=32
    input:
        pc_list=rules.get_pc_list.output,
        geneset="input/{organism}/genesets/{geneset_name}",
    output:
        "out/{organism}/{platform}/pca_ks/"
        "{n_genes}_{scale}_{max_comp}_{var_threshold}/"
        "raw/{geneset_name}.tsv"
    message:
        "Performing GSEA {wildcards.organism} seq \n"
        " Geneset: {wildcards.geneset_name} \n"
        " Number of genes considered: {wildcards.n_genes} \n"
        " Scale of original dataset: {wildcards.scale} \n"
        " Explained variance threshold %: {wildcards.var_threshold} \n"
        " Max PC components for 1 dataset: {wildcards.max_comp} \n"
    log:
        "logs/{organism}/{platform}/ks_genesets/"
        "{n_genes}_{scale}_{max_comp}_{var_threshold}/"
        "{geneset_name}.log"
    conda: "envs/fgsea.yaml"
    shell:
        "Rscript scripts/R/ks_genesets.R {input.pc_list} {input.geneset} {output}"
        " > {log} 2>&1"

rule prepare_pca_ks_result:
    input:
        ks_results=rules.ks_genesets.output,
        gse_df="out/{organism}/{platform}/sm_metadata/gse.tsv"
    output:
        "out/{organism}/{platform}/pca_ks/{n_genes}_{scale}_{max_comp}_{var_threshold}/"
        "prepared/{geneset_name}.tsv"
    message:
        "Preparing results for PCA query. {wildcards.organism} seq \n"
        " Geneset: {wildcards.geneset_name} \n"
        " Number of genes considered: {wildcards.n_genes} \n"
        " Scale of original dataset: {wildcards.scale} \n"
        " Explained variance threshold %: {wildcards.var_threshold} \n"
        " Max PC components for 1 dataset: {wildcards.max_comp} \n"
    log:
        "logs/{organism}/{platform}/prepare_pca_ks_result/"
        "{n_genes}_{scale}_{max_comp}_{var_threshold}/"
        "{geneset_name}.log"
    conda: "envs/r_scripts.yaml"
    shell:
        "Rscript scripts/R/prepare_pca_ks_results.R {input.ks_results} {input.gse_df} {output}"
        " > {log} 2>&1"


# rule mfgsea_genesets:
#     '''
#     Performs ks test on list of PC components. Output ranked by p-val
#     '''
#     resources:
#         mem_ram=32
#     input:
#         pc_list=rules.get_pc_list.output,
#         geneset="input/{organism}/genesets/{geneset_name}",
#     output:
#         "out/{organism}/{platform}/pca_ks/"
#         "{n_genes}_{scale}_{max_comp}_{var_threshold}/"
#         "raw/{geneset_name}.tsv"
#     message:
#         "Performing GSEA {wildcards.organism} seq \n"
#         " Geneset: {wildcards.geneset_name} \n"
#         " Number of genes considered: {wildcards.n_genes} \n"
#         " Scale of original dataset: {wildcards.scale} \n"
#         " Explained variance threshold %: {wildcards.var_threshold} \n"
#         " Max PC components for 1 dataset: {wildcards.max_comp} \n"
#     log:
#         "logs/{organism}/{platform}/ks_genesets/"
#         "{n_genes}_{scale}_{max_comp}_{var_threshold}/"
#         "{geneset_name}.log"
#     conda: "envs/fgsea.yaml"
#     shell:
#         "Rscript scripts/R/mfgsea_genesets.R {input.pc_list} {input.geneset} {output}"
#         " > {log} 2>&1"
#
# rule prepare_pca_mfgsea_result:
#     input:
#         ks_results=rules.ks_genesets.output,
#         gse_df="out/{organism}/{platform}/sm_metadata/gse.tsv"
#     output:
#         "out/{organism}/{platform}/pca_ks/{n_genes}_{scale}_{max_comp}_{var_threshold}/"
#         "prepared/{geneset_name}.tsv"
#     message:
#         "Preparing results for PCA query. {wildcards.organism} seq \n"
#         " Geneset: {wildcards.geneset_name} \n"
#         " Number of genes considered: {wildcards.n_genes} \n"
#         " Scale of original dataset: {wildcards.scale} \n"
#         " Explained variance threshold %: {wildcards.var_threshold} \n"
#         " Max PC components for 1 dataset: {wildcards.max_comp} \n"
#     log:
#         "logs/{organism}/{platform}/prepare_pca_ks_result/"
#         "{n_genes}_{scale}_{max_comp}_{var_threshold}/"
#         "{geneset_name}.log"
#     conda: "envs/r_scripts.yaml"
#     shell:
#         "Rscript scripts/R/prepare_pca_ks_results.R {input.ks_results} {input.gse_df} {output}"
#         " > {log} 2>&1"